repo2docker/buildpacks/python/__init__.py

```diff
@@ -58,26 +82,51 @@ def get_assemble_scripts(self):
                     '${{NB_PYTHON_PREFIX}}/bin/pip install --no-cache-dir -r "{}"'.format(nb_requirements_file)
                 ))
 
-        # install requirements.txt in the kernel env
-        requirements_file = self.binder_path('requirements.txt')
-        if os.path.exists(requirements_file):
+        # install Pipfile.lock, Pipfile, or requirements.txt in the kernel env
+        pipenv = '${KERNEL_PYTHON_PREFIX}/bin/pipenv'
+        pipfile = self.binder_path('Pipfile')
+        pipfile_lock = self.binder_path('Pipfile.lock')
+        if os.path.exists(pipfile_lock) or os.path.exists(pipfile):
+            working_directory = 'binder' if os.path.exists('binder') else '.'
             assemble_scripts.append((
                 '${NB_USER}',
-                'pip install "pip<19" && ' + \
-                '{} install --no-cache-dir -r "{}"'.format(pip, requirements_file)
+                'pip install pipenv'
             ))
-
-        # setup.py exists *and* binder dir is not used
-        if not os.path.exists('binder') and os.path.exists(setup_py):
+            if not os.path.exists(pipfile_lock):
+                assemble_scripts.append((
+                    '${NB_USER}',
+                    '(cd {} && {} lock --python {})'.format(working_directory, pipenv, '${KERNEL_PYTHON_PREFIX}/bin/python')
+                ))
             assemble_scripts.append((
                 '${NB_USER}',
-                '{} install --no-cache-dir .'.format(pip)
+                '(cd {} && {} install --ignore-pipfile --deploy --system --dev --python {})'.format(working_directory, pipenv, '${KERNEL_PYTHON_PREFIX}/bin/python')
             ))
+        else:
+            # KERNEL_PYTHON_PREFIX is the env with the kernel,
+            # whether it's distinct from the notebook or the same.
+            pip = '${KERNEL_PYTHON_PREFIX}/bin/pip'
+            requirements_file = self.binder_path('requirements.txt')
+            setup_py = 'setup.py'
+            if os.path.exists(requirements_file):
+                assemble_scripts.append((
+                    '${NB_USER}',
+                    'pip install "pip<19" && ' + \
```

Comment:
We need to be careful not to reintroduce this `pip` version pin. It has been removed since this PR's branch started. The line was deleted in `master`.